Feature/model functions

mirakst opened the pull request:
DataUtilities.cs: System model population functions, helper functions for building objects for the system models.
SQLDatabase.cs: Query functions

damniko committed:
Created VS project file

damniko committed:
Created all classes (properties and constructors)

Commit:
Added instructions on where to find properties for classes.

Each property has a description of where it comes from and how to query it. Rules are not documented - should be done later. Types of properties have been changed to represent the data type in the column from which they come. This is not true for columns that can contain strings, but we only need numbers.
Can reduce some 'long's to int, or even byte if we care. Should be discussed.

Commit:
Merge branch 'feature/problem-domain-model' into feature/model-functions

Commit:
Fleshed out DB functions and added new ones

Commit:
Added remaining functionality

Did some code refactoring. Started work on an event for modification of validation report and conversion.

Commit:
Rearranged, documented and cleaned up code

All system model building functionality is now in DataUtilities.cs, so SqlDatabase.cs only contains query functions. 

Completed documentation in IDatabaseHandler.cs.

Added comments to functions in DataUtilities.cs.

Removed obsolete 'Using' directives in various files.

Added regions to all model files.

Commit:
Improved documentation, added more function overloading

damniko commented on DashboardBackend/DataUtilities.cs:
```diff
@@ -0,0 +1,323 @@
+            return result;
+        }
+
+        public static List<ValidationTest> GetAfstemninger() => GetAfstemninger(SqlMinDateTime);
```
Mangler dokumentation

mirakst replied:
On it

damniko commented on DashboardBackend/DataUtilities.cs:
```diff
@@ -0,0 +1,323 @@
+        public static List<LogMessage> GetLogMessages(int ExecutionId, DateTime minDate)
+        {
+            List<LoggingEntry> queryResult;
+            if (ExecutionId > 0)
```
overvej at kaste en exception i stedet, for det er ikke meningen at man skal smide et negativt ID ind - og så kan man ikke forvente at det kører som det skal

damniko commented on DashboardBackend/DataUtilities.cs:
```diff
@@ -0,0 +1,323 @@
+
+            foreach (LoggingEntry item in queryResult)
+            {
+                result.Add(new LogMessage(item.LogMessage, DataUtilities.GetLogMessageType(item), (int)item.ContextId, (DateTime)item.Created));
```
personligt vil jeg hellere scrolle vertikalt end horisontalt - så det her ville jeg selv bryde op i mindre dele, f.eks.
```
LogMessageType type = DataUtilities.GetLogMessageType(item);
int contextId = int.Parse(item.ContextId);
DateTime created = item.Created as DateTime; (eller whatever)
LogMessage message = new(log.LogMessage, type, contextId, created);
result.Add(messsage);
```
Så er der også meget bedre muligheder for input-validering

damniko commented on DashboardBackend/DataUtilities.cs:
```diff
@@ -0,0 +1,323 @@
+        {
+            List<HealthReportEntry> queryResult = DatabaseHandler.QueryHealthReport();
+
+            HealthReport result = BuildHealthReport(queryResult);
```
det er nok lidt overkill med mellemrum mellem hver linje - men det er måske bare mig der synes det

damniko commented on DashboardBackend/DataUtilities.cs:
```diff
@@ -0,0 +1,323 @@
+        /// </summary>
+        /// <param name="minDate">The minimum DateTime for the query results.</param>
+        /// <returns>A list of CPU load readings.</returns>
+        public static List<CpuLoad> GetCpuReadings(DateTime minDate)
```
Kunne man omstrukturere alt der her health report halløj til noget i stil med:

1. Query alt i Health Report (sorteret efter tid)
2. Gå listen igennem fra toppen af og kør de her funktioner ud fra en switch med REPORT_KEY (Network, Memory, CPU, etc.)

Så kan man nøjes med at query det hele én gang

damniko commented on DashboardBackend/DataUtilities.cs:
```diff
@@ -0,0 +1,323 @@
+        /// <exception cref="ArgumentException">Thrown if the parameter passed is not a legal log message type.</exception>
+        public static LogMessageType GetLogMessageType(LoggingEntry entry)
+        {
+            if (entry.LogMessage.StartsWith("Afstemning"))
```
Skal i stedet være:
```
if (entry.LogMessage.Contains("Afstemning") || entry.LogMessage.StartsWith("Check -"))
{ ... }
```

damniko commented on DashboardBackend/DataUtilities.cs:
```diff
@@ -0,0 +1,323 @@
+        /// <param name="entry">A single entry from the [LOGGING] table in the state database.</param>
+        /// <returns>A log message type besed on the enum in the log message class.</returns>
+        /// <exception cref="ArgumentException">Thrown if the parameter passed is not a legal log message type.</exception>
+        public static LogMessageType GetLogMessageType(LoggingEntry entry)
```
Den her funktion kan vel også fjernes ved at lave en dictionary fra strenge til LogMessageTypes, så:
`LogMessageType type = TypeDict[logMessage.Type];`

damniko commented on DashboardBackend/Database/SqlDatabase.cs:
```diff
@@ -12,51 +9,109 @@ public SqlDatabase() : base()
+            using NetcompanyDbContext db = new();
+            var queryResult = db.Executions
+                                .Where(e => e.Created > minDate)
+                                .OrderBy(e => e.Created).ToList();
```
Enten ryk `ToList()` ned til `return queryResult.ToList()` eller lad være med at have queryResult

damniko replied:
... og hvis den fortsat skal være der, så sæt den i det mindste ned på næste linje!

damniko commented on Model/Execution.cs:
```diff
@@ -0,0 +1,52 @@
+        public DateTime EndTime { get; private set; } //DateTime.Now when an execution is registered as done (from log?).
+        public TimeSpan Runtime { get; private set; } //EndTime.Subtract(StartTime)
+        public int RowsReadTotal { get; set; } //OnExecutionFinished, for each manager, RowsReadTotal += RowsRead.
+        public int RowsWrittenTotal { get; set; } //OnExecutionFinished, for each manager, RowsWrittenTotal += RowsWritten.
```
er vi interesseret i det her for en execution?

damniko replied:
er det ikke mere for hver manager?

damniko commented on Model/ValidationReport.cs:
```diff
@@ -0,0 +1,42 @@
+            set
+            {
+                _validationTests.AddRange(value);
+                OnModified(); // Raises event when property is set. Possible parameters include = List content, list count etc.
```
lidt usikker på hvad meningen er med at have den event herinde

damniko reviewed (approved):
ser funktionelt ud, men har nogle mindre kommentarer

Commit:
Added more documentation, refactored code

Created a function to get performance readings from one database query.
Added missing documentation.
Removed leftover code.
